File: 10_flutter_and_native_samples/ios/Runner/NativeComputationChannel.swift
import Flutter
import Foundation

/// Runs a delayed native task on a background queue and returns the
/// result to Flutter asynchronously on the main thread.
final class NativeComputationChannel {
  static let channelName = "examples.flutter_native_calls/async_task"

  private let channel: FlutterMethodChannel

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
  }

  func register() {
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
  }

  func unregister() {
    channel.setMethodCallHandler(nil)
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "runDelayedTask":
      runDelayedTask(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func runDelayedTask(result: @escaping FlutterResult) {
    // Native work: sleep a random number of seconds, then reply on the main thread.
    DispatchQueue.global(qos: .userInitiated).async {
      let delaySeconds = Int.random(in: 1...10)
      let start = DispatchTime.now()
      Thread.sleep(forTimeInterval: TimeInterval(delaySeconds))
      let durationMs = Int(
        (DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000
      )

      DispatchQueue.main.async {
        result([
          "result": "Native delay completed after \(delaySeconds) s",
          "durationMs": durationMs,
        ])
      }
    }
  }
}

File: 10_flutter_and_native_samples/android/app/src/main/kotlin/com/example/flutter_and_native_samples/NativeComputationChannel.kt
package com.example.flutter_and_native_samples

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.random.Random

/**
 * Runs a delayed native task on a background thread and returns the
 * result to Flutter asynchronously on the main thread.
 */
class NativeComputationChannel(
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler {
    private val channel =
        MethodChannel(messenger, "examples.flutter_native_calls/async_task")

    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    fun register() {
        channel.setMethodCallHandler(this)
    }

    fun unregister() {
        channel.setMethodCallHandler(null)
        executor.shutdownNow()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "runDelayedTask" -> runDelayedTask(result)
            else -> result.notImplemented()
        }
    }

    private fun runDelayedTask(result: MethodChannel.Result) {
        // Native work: sleep a random number of seconds, then reply on the main thread.
        executor.execute {
            val delaySeconds = Random.nextInt(1, 11)
            val start = System.nanoTime()
            try {
                Thread.sleep(delaySeconds * 1_000L)
            } catch (_: InterruptedException) {
                mainHandler.post {
                    result.error(
                        "unavailable",
                        "Native delayed task was interrupted.",
                        null,
                    )
                }
                return@execute
            }
            val durationMs = ((System.nanoTime() - start) / 1_000_000).toInt()

            mainHandler.post {
                result.success(
                    mapOf(
                        "result" to "Native delay completed after $delaySeconds s",
                        "durationMs" to durationMs,
                    ),
                )
            }
        }
    }
}
